src/aleph/vm/orchestrator/payment.py

```diff
@@ -0,0 +1,160 @@
+    if is_address(hex_address):
+        return HexAddress(HexStr(hex_address))
+    else:
+        raise ValueError("Unknown format {}, attempted to normalize to {}".format(value, hex_address))
+
+
+def get_required_balance(executions: Iterable[VmExecution]) -> Decimal:
+    """Get the balance required for the resources of the user from the messages and the pricing aggregate."""
+    balance = Decimal(0)
+    for execution in executions:
+        balance += compute_execution_hold_cost(execution)
+
+    return Decimal(balance)
+
+
+def compute_execution_hold_cost(execution: VmExecution) -> Decimal:
+    compute_unit_cost = 200 if execution.persistent else 2000
+
+    compute_units_required = _get_nb_compute_units(execution)
+    compute_unit_multiplier = _get_compute_unit_multiplier(execution)
+
+    compute_unit_price = Decimal(compute_units_required) * compute_unit_multiplier * compute_unit_cost
+    price = compute_unit_price + _get_additional_storage_hold_price(execution)
+    return Decimal(price)
+
+
+def _get_additional_storage_hold_price(execution: VmExecution) -> Decimal:
+    nb_compute_units = execution.vm.hardware_resources.vcpus
+    free_storage_per_compute_unit = 2 * GiB if not execution.persistent else 20 * GiB
+
+    total_volume_size = _get_execution_storage_size(execution)
+    additional_storage = max(total_volume_size - (free_storage_per_compute_unit * nb_compute_units), 0)
+    price = Decimal(additional_storage) / 20 / MiB
+    return price
+
+
+def _get_nb_compute_units(execution: VmExecution) -> int:
+    cpu = execution.vm.hardware_resources.vcpus
+    memory = math.ceil(execution.vm.hardware_resources.memory / 2048)
+    nb_compute_units = cpu if cpu >= memory else memory
+    return nb_compute_units
+
+
+def _get_compute_unit_multiplier(execution: VmExecution) -> int:
+    compute_unit_multiplier = 1
+    if not execution.persistent and execution.message.environment.internet:
+        compute_unit_multiplier += 1
+    return compute_unit_multiplier
+
+
+def _get_execution_storage_size(execution: VmExecution) -> int:
+    size = 0
+
+    if execution.is_instance:
+        size += execution.message.rootfs.size_mib * MiB
+    elif execution.is_program:
+        if isinstance(execution.resources, AlephProgramResources):
+            size += get_path_size(execution.resources.code_path)
+            if execution.resources.data_path:
+                size += get_path_size(execution.resources.data_path)
+
+    for volume in execution.resources.volumes:
+        size += get_path_size(volume.path_on_host)
+
+    return size
+
+
+def get_required_flow(executions: Iterable[VmExecution]) -> Decimal:
+    """Compute the flow required for the resources of the user from the messages and the pricing aggregate"""
+    flow = Decimal(0)
+    for execution in executions:
+        flow += compute_execution_flow_cost(execution)
+
+    return Decimal(flow)
```

Comment:
Looks like a duplicate to `get_required_balance()`. Maybe fuse them into one function?

Reply:
It's a different method, for a different purpose, and I don't think that merge will be good for code understanding.